### PR TITLE
Speedup rt0

### PR DESCRIPTION
**Pull request by alessiofumagalli**
small branch to speed up the rt0 solver.
in particular, apart from local changes, some geometrical mapping are computed once and stored in the data structure.

### CODE REVIEW

**Comment by alessiofumagalli** [comment]
NOTE: in a time dependent example after the first run the cost at least half

File: src/porepy/numerics/fem/rt0.py [hunk 1]
```diff
@@ -94,27 +94,37 @@ def discretize(self, g, data):
...
+        elif g.dim == 2:
+            inv_matrix = self._inv_matrix_2d
+        elif g.dim == 3:
+            inv_matrix = self._inv_matrix_3d
+
+        # compute the oppiside node per face
+        self._compute_cell_face_to_opposide_node(g, data)
+        cell_face_to_opposide_node = data[self.cell_face_to_opposide_node]
+
         for c in np.arange(g.num_cells):
```
Comment (rbe051) [inline src/porepy/numerics/fem/rt0.py]: I guess if you want even more speedup, these loops should be possible to vectorize? Of cause not a must for this pull request

File: src/porepy/numerics/fem/rt0.py [hunk 2]
```diff
@@ -221,41 +240,54 @@ def massHdiv(K, c_volume, coord, sign, dim, HB):
...
 
-        return np.dot(C.T, np.dot(N.T, np.dot(HB, np.dot(inv_K, np.dot(N, C)))))
+        return np.dot(C.T, np.dot(N.T, np.dot(HB, np.dot(inv_K_exp, np.dot(N, C)))))
 
-    @staticmethod
-    def opposite_side_node(face_nodes, nodes, faces_loc):
-        """
-        Given a face return the node on the opposite side, typical request of a Raviart-Thomas
-        approximation. This function is mainly for internal use.
+    def _compute_cell_face_to_opposide_node(self, g, data):
```
Comment (rbe051) [inline src/porepy/numerics/fem/rt0.py]: Should it be possible to recompute by passing an argument? I.e. def _compute_cell_face_to_opposide_node(self, g, data, recompute=False)

Comment (rbe051) [reply src/porepy/numerics/fem/rt0.py]: Can you add a short documentation at the beginning of the function? (possible heavily inspired by the documentation from the deleted function opposite_side_node)

File: src/porepy/numerics/vem/dual_elliptic.py [hunk 3]
```diff
@@ -559,3 +559,75 @@ def extract_pressure(self, g, solution_array, data):
...
+
+        Parameters
+        ----------
+        K : the matrix to be inverted 2x2
+
+        Return
+        ------
+        The inverted matrix 2x2
+        """
+        den = K[0, 0] * K[1, 1] - K[0, 1] * K[0, 1]
```
Comment (rbe051) [inline src/porepy/numerics/vem/dual_elliptic.py]: den -> det ?

File: src/porepy/numerics/fem/rt0.py (hunk 1 above)
Comment (alessiofumagalli) [reply src/porepy/numerics/fem/rt0.py]: that would be awesome, no idea how to do it :(

File: src/porepy/numerics/vem/dual_elliptic.py (hunk 3 above)
Comment (alessiofumagalli) [reply src/porepy/numerics/vem/dual_elliptic.py]: yes, it should be det. thanks

File: src/porepy/numerics/fem/rt0.py [hunk 4]
```diff
@@ -12,14 +12,14 @@
...
 
-
 # Module-wide logger
 logger = logging.getLogger(__name__)
 
 
 class RT0(pp.numerics.vem.dual_elliptic.DualElliptic):
     def __init__(self, keyword):
         super(RT0, self).__init__(keyword, "RT0")
+        self.cell_face_to_opposide_node = "rt0_class_cell_face_to_opposide_node"
```
Comment (keileg) [inline src/porepy/numerics/fem/rt0.py]: Documentation?

File: src/porepy/numerics/fem/rt0.py (hunk 1 above)
Comment (keileg) [reply src/porepy/numerics/fem/rt0.py]: Have you tried to numba it? That will go really fast, but it has a startup cost that often is a second or so. Or you can try cython.

See numercis.fv.fvutils.invert_diagonal_blocks() and associated functions for examples of both.

File: src/porepy/numerics/fem/rt0.py [hunk 5]
```diff
@@ -94,27 +94,37 @@ def discretize(self, g, data):
...
+        cell_face_to_opposide_node = data[self.cell_face_to_opposide_node]
+
         for c in np.arange(g.num_cells):
             # For the current cell retrieve its faces
             loc = slice(g.cell_faces.indptr[c], g.cell_faces.indptr[c + 1])
             faces_loc = faces[loc]
 
-            # find the opposite node id for each face
-            node = RT0.opposite_side_node(g.face_nodes, nodes, faces_loc)
+            # get the opposite node id for each face
```
Comment (keileg) [inline src/porepy/numerics/fem/rt0.py]: Precise comments. Very good!

File: src/porepy/numerics/fem/rt0.py (hunk 2 above)
Comment (keileg) [reply src/porepy/numerics/fem/rt0.py]: I agree, comments are necessary.

Comment (alessiofumagalli) [reply src/porepy/numerics/fem/rt0.py]: okay, I've added comments now

**Comment by alessiofumagalli** [comment]
okay, a new version is updated. let me know what do you think

**Comment by alessiofumagalli** [comment]
@keileg

**Pull request closed without merging**